File: 백준/레벨Gold IV/2179. 비슷한 단어/비슷한 단어.py
from sys import stdin

def solution():    
    read = stdin.read().splitlines()
    
    N = int(read[0].rstrip())
    
    # 사전 순 정렬
    words = [[read[i], i] for i in range(1, len(read))]
    words.sort(key=lambda x: x[0])

    # 인접 글자 간 LCP 처리
    max_len = 0
    result = set()
    for i in range(len(words) - 1):
        pre, pre_idx = words[i]
        nxt, nxt_idx = words[i + 1]
        size = min(len(pre), len(nxt))

        # LCP 길이 확인
        cur_len = 0
        for j in range(size):
            if pre[j] == nxt[j]:
                cur_len += 1
            else:
                break
        
        # LCP 갱신
        if max_len < cur_len:
            lcp = pre[:cur_len]
            max_len = cur_len
            result = {(pre_idx, lcp), (nxt_idx, lcp)}
        elif max_len == cur_len:
            lcp = pre[:cur_len]
            result.add((pre_idx, lcp))
            result.add((nxt_idx, lcp))

    # 여러 개일 때에는 입력되는 순서대로
    idx_list = sorted(result)
    idx1, common = idx_list[0]

    idx2 = -1
    for i in range(1, len(idx_list)):
        if idx_list[i][1] == common:
            idx2 = idx_list[i][0]
            break

    print(read[idx1])
    print(read[idx2])

solution()

File: 백준/레벨Gold IV/2179. 비슷한 단어/비슷한 단어.cc
#include <bits/stdc++.h>
using namespace std;

struct Word {
    string text;
    int index;

    Word(string t, int i) : text(t), index(i) {}
};

void solution() {
    int N;
    cin >> N;
    cin.ignore();

    vector<string> read(N + 1);
    for (int i = 1; i <= N; i++) {
        getline(cin, read[i]);
    }

    // 사전 순 정렬
    vector<Word> words;
    for (int i = 1; i <= N; i++) {
        words.emplace_back(read[i], i);
    }
    sort(words.begin(), words.end(), [](const Word& a, const Word& b) {
        return a.text < b.text;
    });

    // 인접 글자 간 LCP 처리
    int maxLen = 0;
    set<pair<int, string>> result;

    for (int i = 0; i < words.size() - 1; i++) {
        string pre = words[i].text;
        int preIdx = words[i].index;

        string nxt = words[i + 1].text;
        int nxtIdx = words[i + 1].index;

        int size = min(pre.size(), nxt.size());
        int curLen = 0;

        for (int j = 0; j < size; j++) {
            if (pre[j] == nxt[j]) curLen++;
            else break;
        }

        // LCP 갱신
        if (maxLen < curLen) {
            string lcp = pre.substr(0, curLen);
            maxLen = curLen;
            result.clear();
            result.emplace(preIdx, lcp);
            result.emplace(nxtIdx, lcp);
        } else if (maxLen == curLen) {
            string lcp = pre.substr(0, curLen);
            result.emplace(preIdx, lcp);
            result.emplace(nxtIdx, lcp);
        }
    }

    // 여러 개일 때에는 입력되는 순서대로(set은 정렬을 보장)
    vector<pair<int, string>> idxList(result.begin(), result.end());
    int idx1 = idxList[0].first;
    string common = idxList[0].second;

    int idx2 = -1;
    for (size_t i = 1; i < idxList.size(); i++) {
        if (idxList[i].second == common) {
            idx2 = idxList[i].first;
            break;
        }
    }

    cout << read[idx1] << '\n';
    cout << read[idx2] << '\n';
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    solution();
}